proposals/0000-fine-grained-unused-warnings.rst

```diff
@@ -0,0 +1,205 @@
+actually that ``foo`` is unused, but rather that all of ``foo``'s usage sites
+are in unused bindings.
+
+Furthermore, while it's useful to see such a warning for CI, to immediately
+know all of the bindings that have to be removed to eliminate the warnings,
+during development, these additional warnings take up a lot of valuable
+screen real estate in ghcid.
+
+Proposed Change Specification
+-----------------------------
+The overarching idea is to separate warnings about names that are genuinely
+unused, and warnings that are only "transitively unused" (or "indirectly unused"), i.e. only used by
+other (transitively) unused bindings.
+
+The warnings about transitively unused names will be reported if a new flag, `-freport-indirectly-unsed-bindings` is enabled (by default, it is) *and* the appropriate existing warning flags about unused bindings is enabled, i.e. one of
+
+* ``-Wunused-top-binds``
+* ``-Wunused-local-binds``
+* ``-Wunused-pattern-binds``
+* ``-Wunused-binds``
+* ``-Wunused-foralls``
+* ``-Wunused-matches``
+* ``-Wunused-imports``
+* ``-Wunused-type-patterns``
+
+In each of these cases, if a name is not used by anything, the same warning will be produced that has been produced prior to this proposal.
+
+For example, in the following code
+
+::
+
+  module Foo ()
+
+  import Data.List as L
+
+  foo = L.intercalate
+
+``intercalate`` is transitively unused, since it is only used by ``foo``, which is unused. The warning that's produced will be
+
+::
+
+  Foo.hs:5:1: warning: [-Wunused-top-binds]
+      Defined but not used: ‘foo’
+  
+  Foo.hs:3:1: warning: [-Wunused-imports, -freport-indirectly-used-bindings]:
+      The import of `Data.List' is only used by the following unused binding: ‘foo’
+        except perhaps to import instances from `Data.List'
+      To import instances alone, use: import Data.List()
+
+The warning about transitively unused bindings only appears if both the warning flag
+for the unused binding (in this case ``-Wunused-top-binds``) and the the one for
+the transitively unused binding itself (in this case ``-Wunused-imports``) are
+enabled.
+
+If ``foo`` is only used in a situation like
+
+::
+
+  bar = baz + 2
+    where baz = foo * 2
+	
+and bar is unused, a question that arises is whether the warning about ``foo`` should reference ``bar``, or ``baz``.
+
+In this particular case, since there would be no warning about ``baz`` being unused, it seem more helpful if the warning references ``bar``.
+
+In general, a warning should only reference other bindings about which there is an "unused" warning.
+
+In practice, that typically means warnings about transitively unused top-level bindings will tend to reference other top-level bindings, and warnings about transitively unused
+local bindings will reference other local bindings. However, if ``baz`` were unused, it would be referenced in the warning instead, since using ``bar`` in that case would not get rid of the warning about ``foo``.
```

Comment:
I'm finding this section a bit hard to follow. It would be nice to nail down the specification a bit more precisely: instead of discussing hypothetical designs and what "should" be the case, can we give a clear definition of which bindings are included in the list? (By all means unpack the meaning of the definition with examples and discussion of why it is the right design.)

Reply:
I've rewritten the section, let me know what you think